metrics/resource_view.go

```diff
@@ -0,0 +1,295 @@
+	"go.opencensus.io/tag"
+)
+
+type storedViews struct {
+	views []*view.View
+	lock  sync.Mutex
+}
+
+type meterExporter struct {
+	m view.Meter    // NOTE: DO NOT RETURN THIS DIRECTLY; the view.Meter will not work for the empty Resource
+	o stats.Options // Cache the option to reduce allocations
+	e view.Exporter
+}
+
+// ResourceExporterFactory provides a hook for producing separate view.Exporters
+// for each observed Resource. This is needed because OpenCensus support for
+// Resources is a bit tacked-on rather than being a first-class component like
+// Tags are.
+type ResourceExporterFactory func(*resource.Resource) (view.Exporter, error)
+type meters struct {
+	meters  map[string]*meterExporter
+	factory ResourceExporterFactory
+	// Cache of Resource pointers from metricskey to Meters, to avoid
+	// unnecessary stringify operations
+	resourceToKey map[*resource.Resource]string
+	lock          sync.Mutex
+}
+
+// Lock regime: lock allMeters before resourceViews. The critical path is in
+// optionForResource, which must lock allMeters, but only needs to lock
+// resourceViews if a new meter needs to be created.
+var resourceViews storedViews = storedViews{}
+var allMeters meters = meters{
+	meters:        map[string]*meterExporter{"": &defaultMeter},
+	resourceToKey: map[*resource.Resource]string{nil: ""},
+}
+
+// RegisterResourceView is similar to view.Register(), except that it will
+// register the view across all Resources tracked by the system, rather than
+// simply the default view.
+func RegisterResourceView(views ...*view.View) error {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	errors := make([]error, 0, len(allMeters.meters))
+
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+	for _, meter := range allMeters.meters {
+		if err := meter.m.Register(views...); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return errors[0] // The first error is as good as any
+	}
+
+	resourceViews.views = append(resourceViews.views, views...)
+	return nil
+}
+
+func UnregisterResourceView(views ...*view.View) {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		meter.m.Unregister(views...)
+	}
+
+	j := 0
+	for _, view := range resourceViews.views {
+		toRemove := false
+		for _, viewToRemove := range views {
+			if view == viewToRemove {
+				toRemove = true
+				break
+			}
+		}
+		if !toRemove {
+			resourceViews.views[j] = view
+			j++
+		}
+	}
+	resourceViews.views = resourceViews.views[:j]
+}
+
+func setFactory(f func(*resource.Resource) (view.Exporter, error)) error {
+	if f == nil {
+		return errors.New("do not setFactory(nil)")
+	}
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	allMeters.factory = f
+
+	errs := make([]error, len(allMeters.meters))
+
+	for r, meter := range allMeters.meters {
+		e, err := f(resourceFromString(r))
```

Comment:
Any reason to continue after an error?

Reply:
To Unregister as many old exporters as possible, rather than to leave the system in a more-dangling state?

Reply:
Given we only return the first error, `errs` is probably overkill. Just keep the last err (since we don't care which one we return).